### PR TITLE
Add generic lambda module.

### PR DESCRIPTION
**Pull request by MancunianSam**
We now have to create a few new lambdas and I don't want to have to use
the old system of having a file with a count for each lambda so I've
created this generic lambda module.

Outside of the standard configuration like name and memory size, you can
configure the following:
* SQS queue permissions
* API Gateway lambda invocation permission
* EFS config (Hopefully, we won't need these)
* Plaintext environment variables
* Encrypted environment variables.

The module creates a role with a name that is passed in as a variable.
It also creates a policy from a name and policy string which is passed
in as a variable.

I'm not 100% sure this is the best approach, it might be better to
create the roles and policies outside the lambda module and pass them in
but that will make the root files very verbose so I think this is better
for now. It will be an easy enough change if we need to.

### CODE REVIEW

File: generic_lambda/main.tf [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+resource "aws_lambda_permission" "export_api_lambda_permissions" {
+  for_each      = var.lambda_invoke_permissions
+  statement_id  = "AllowExecutionFrom${title(split(".", each.key)[0])}"
+  action        = "lambda:InvokeFunction"
+  function_name = aws_lambda_function.lambda_function.function_name
+  principal     = each.key
+  source_arn    = each.value
+}
+
```
Comment (TomJKing) [inline generic_lambda/main.tf:72]: Unnecessary line break

File: generic_lambda/main.tf [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  filename      = startswith(var.runtime, "java") ? "${path.module}/functions/generic.jar" : "${path.module}/functions/generic.zip"
+  timeout       = var.timeout_seconds
+  memory_size   = var.memory_size
+
+  ephemeral_storage {
+    size = var.storage_size
+  }
+
+  reserved_concurrent_executions = var.reserved_concurrency
+  tags                           = var.common_tags
```
Comment (TomJKing) [inline generic_lambda/main.tf:15]: Maybe rename this var as just `tags`given they're just the tags for the lambda now, and not a set of common ones

**Review by TomJKing — COMMENTED** [review]
Looks good. Thanks for doing this. Couple of very minor comments

File: generic_lambda/main.tf (hunk 2 above)
Comment (MancunianSam) [reply generic_lambda/main.tf:15]: Good point, I've changed it.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks

**Pull request merged**